src/Symfony/Component/Security/Core/Authentication/Provider/UserAuthenticationProvider.php

```diff
@@ -93,7 +93,13 @@ public function authenticate(TokenInterface $token)
             throw $e;
         }
 
-        $authenticatedToken = new UsernamePasswordToken($user, $token->getCredentials(), $this->providerKey, $this->getRoles($user, $token));
+        $previousToken = null;
+
+        if ($token instanceof UsernamePasswordToken) {
+            $previousToken = $token->getPreviousToken();
+        }
+
+        $authenticatedToken = new UsernamePasswordToken($user, $token->getCredentials(), $this->providerKey, $this->getRoles($user, $token), $previousToken);
```

Comment:
What about using a `SwitchUserToken()` to decorate the previous one, that would help identifying it elsewhere.

Reply:
This is now a thing.